fixes #1994 - set vardir for puppet 3, use --configprint

domcleal opened the pull request:
When running Puppet 3 as a non-root user (i.e. foreman), both --confdir and
--vardir have to be specified as per Puppet #16637, else SSL settings dependent
on vardir will fail.

This adds a new puppetvardir setting used with puppetconfdir, and supports
puppetconfdir being a file or directory due to historic settings.yaml examples.

It now shells out and uses `puppet master --configprint` to retrieve settings
to avoid changes in Puppet settings initialization logic across versions (e.g.
run_mode on Puppet 3).

domcleal commented:
Updated branch with a change to call `--configprint foo,bar,baz` in a single shot, which runs much faster as discussed with @ohadlevy.

ohadlevy commented on lib/foreman/default_settings/loader.rb:
```diff
@@ -20,13 +22,55 @@ def create opts
         end
 
-        def load(reset=false)
+        def find_puppetmaster
```
I think this code should be in a seperate class

how about adding a class called /lib/puppet_setting.rb ?

ohadlevy replied:
it would be later on trivial to change the internal class behavior if there would be a better way in the future?

domcleal replied:
sounds sensible, will do

ohadlevy commented on test/lib/foreman/util_test.rb:
```diff
@@ -0,0 +1,40 @@
+require 'test_helper'
+require 'foreman/util'
+
```
those tests are actually useful to the proxy too (I'm guessing that's where you took the which method).

domcleal replied:
@ohadlevy yep, shall I submit them there too?

ohadlevy replied:
:+1:

domcleal commented:
@ohadlevy updated with separate PuppetSetting

Commit:
fixes #1994 - set vardir for puppet 3, use --configprint

When running Puppet 3 as a non-root user (i.e. foreman), both --confdir and
--vardir have to be specified as per Puppet #16637, else SSL settings dependent
on vardir will fail.

This adds a new puppetvardir setting used with puppetconfdir, and supports
puppetconfdir being a file or directory due to historic settings.yaml examples.

It now shells out and uses `puppet master --configprint` to retrieve settings
to avoid changes in Puppet settings initialization logic across versions (e.g.
run_mode on Puppet 3).

ohadlevy commented:
merged - thanks!
